### DISCUSSION TITLE
how to use the ldap plugin

### DISCUSSION THREAD

**Discussion by liangliwu** [apache/apisix#5929] [Q&A]
Why don't the properties of the ldap plugin include bind dn and bind dn's paasword. I tried to integrate the ldap plugin with open ldap, but it failed.

**Comment by juzhiyuan** [comment]
Hi @liangliwu, here is the LDAP-auth plugin doc: https://apisix.apache.org/docs/apisix/plugins/ldap-auth/

And I noticed the password is extracted from Authorization header, does it make sense to you? https://github.com/apache/apisix/blob/master/apisix/plugins/ldap-auth.lua#L132-L141

**Reply by liangliwu** [reply]
I grabbed the packet via tcpdump and saw that the ldap server responded with bind success. but apisix still returned {"message": "Invalid API key in request"} for the customer.

**Reply by juzhiyuan** [reply]
Hi @liangliwu, could you please share your Route and Consumer's configuration here?

**Reply by liangliwu** [reply]
the consumer‘s configuration like follow:
{
  "username": "test",
  "desc": "test",
  "plugins": {
    "ldap-auth": {
      "disable": false,
      "user_dn": "cn=test,ou=users,ou=macaopso,dc=abc,dc=com,dc=mo"
    }
  }
}

the route's configuration like follow:
{
  "uris": [
    "/webportal/*",
    "/portal/*"
  ],
  "name": "wwwtest",
  "methods": [
    "GET",
    "POST",
    "PUT",
    "DELETE",
    "PATCH",
    "HEAD",
    "OPTIONS",
    "CONNECT",
    "TRACE"
  ],
  "host": "vf.abc.com",
  "plugins": {
    "ldap-auth": {
      "base_dn": "ou=macaopso,ou=users,dc=abc,dc=com,dc=mo",
      "disable": false,
      "ldap_uri": "xxx.xxx.xxx.xxx:389",
      "uid": "cn"
    }
  },
  "service_id": "386663780244784120",
  "labels": {
    "API_VERSION": "1.0.0"
  },
  "status": 1
}

**Reply by juzhiyuan** [reply]
Hi @leslie-tsang, could you please have a check?

**Reply by kostyanp95** [reply]
> I grabbed the packet via tcpdump and saw that the ldap server responded with bind success. but apisix still returned {"message": "Invalid API key in request"} for the customer.

Did you manage to solve this problem?
I have exactly the same. Intercepted the packet via WireShark...

### Request to gateway route:
```
curl -i -u ldap_id_nexus http://127.0.0.1:9080/service1
Enter host password for user 'ldap_id_nexus':
HTTP/1.1 401 Unauthorized
Date: Fri, 13 May 2022 13:55:56 GMT
Content-Type: text/plain; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Server: APISIX/2.13.0

{"message":"Invalid API key in request"}
```

### Response from ldap in WireShark:
![image](https://user-images.githubusercontent.com/51831069/168303260-7a7b6b68-221f-4f52-ade9-75ff2636b381.png)

### Error Search:
I looked for an error in the APISIX repository, found it only in the **key-auth** plugin... How can this be?
![image](https://user-images.githubusercontent.com/51831069/168306473-08932d7c-c739-4402-87e2-a1a5d2084c34.png)

The text of this error is not in the lua script of the ldap-auth plugin...
![image](https://user-images.githubusercontent.com/51831069/168312076-7999e790-fb6a-4426-b370-f8738b1835c1.png)

### Consumer and Route Configuration:

- **Consumer**:
```
{
  "username": "Test",
  "plugins": {
    "ldap-auth": {
      "user_dn": "cn=ldap_id_nexus,ou=TechAccount,dc=city,dc=company,dc=zone"
    }
  }
}
```

- **Route**:
```
{
  "uri": "/service1",
  "name": "TEST",
  "methods": [
    "GET"
  ],
  "plugins": {
    "ldap-auth": {
      "base_dn": "OU=TechAccount,DC=city,DC=company,DC=zone",
      "disable": false,
      "ldap_uri": "ldap.server"
    }
  },
  "upstream": {
    "nodes": [
      {
        "host": "xxx.xx.xx.xxx",
        "port": 80,
        "weight": 1
      }
    ],
    "timeout": {
      "connect": 6,
      "send": 6,
      "read": 6
    },
    "type": "roundrobin",
    "scheme": "http",
    "pass_host": "pass",
    "keepalive_pool": {
      "idle_timeout": 60,
      "requests": 1000,
      "size": 320
    }
  },
  "status": 1
}
```

Thank you so much for any feedback!🙏🏼

**Reply by Chever-John** [reply]
Hi, Can you check the APISIX logs? to see if the plugin was successfully opened?

**Reply by moonming** [reply]
@tzssangglass  please take a look,thx

**Reply by tzssangglass** [reply]
Hi @kostyanp95 , you can enable: https://github.com/apache/apisix/blob/master/docs/en/latest/architecture-design/debug-mode.md#basic-debug-mode to check if the request has gone through the key-auth plugin.

**Reply by kostyanp95** [reply]
@jp-gouin, hi! 
Also, I noticed that the `ldap-auth.lua` script code here on github differs from the actual one that lies in the docker container.

I used the command `docker exec -it apisix bash` and `cat apisix/plugins/ldap-auth.lua` to view the code.

### Comparison of ldap-auth.lua file on GitHub and in docker container:
On the left is a file copied from [GitHub](https://github.com/apache/apisix/blob/382e710e99341cf693ed5a5aa226657deed0bdc5/apisix/plugins/ldap-auth.lua), on the right is a file from the docker container apisix.

![image](https://user-images.githubusercontent.com/51831069/168804114-86f39e8c-2477-4e95-87c0-0b519daac6e3.png)

As we can see, in the script from the docker container on the right on line 155, this strange message `Invalid API key in the request` was found.

Did something go wrong? :)

**Reply by tzssangglass** [reply]
ok, I know, maybe cause by this PR: https://github.com/apache/apisix/pull/6707/files#diff-8e5b15fc718c151b0f2b2fbf464e510f31c3e0bf31a350aa2f5ca01349f4a2f6

I think your problem is a consumer mismatch

```
"user_dn": "cn=ldap_id_nexus,ou=TechAccount,dc=city,dc=company,dc=zone"
```

and 

```
"base_dn": "OU=TechAccount,DC=city,DC=company,DC=zone",
```

it looks like they are not case-sensitive, can you try:

```
"base_dn": "ou=TechAccount,dc=city,dc=company,dc=zone",
```

**Reply by kostyanp95** [reply]
> ok, I know, maybe cause by this PR: https://github.com/apache/apisix/pull/6707/files#diff-8e5b15fc718c151b0f2b2fbf464e510f31c3e0bf31a350aa2f5ca01349f4a2f6
> 
> I think your problem is a consumer mismatch
> 
> ```
> "user_dn": "cn=ldap_id_nexus,ou=TechAccount,dc=city,dc=company,dc=zone"
> ```
> 
> and
> 
> ```
> "base_dn": "OU=TechAccount,DC=city,DC=company,DC=zone",
> ```
> 
> it looks like they are not case-sensitive, can you try:
> 
> ```
> "base_dn": "ou=TechAccount,dc=city,dc=company,dc=zone",
> ```

Indeed, I changed user_dn to 
```"user_dn": "cn=ldap_id_nexus,OU=TechAccount,DC=city,DC=company,DC=zone"```
@liangliwu, that is, `cn`, `ou` and `dc` are case sensitive. And it worked!

Also, as it turns out, `cn` cannot contain whitespace characters, because plugin deletes those. It is normal? @jp-gouin
([ldap-auth.lua](https://github.com/apache/apisix/blob/6ebe02797fb564c84b30df1865adef7f473880bf/apisix/plugins/ldap-auth.lua#L116))

Anyway, thanks for your help everyone!

**Comment by kostyanp95** [comment]
Do we mark the issue as resolved?

**Comment by qyk1995** [comment]
@kostyanp95     help me, I  meet same question.  https://github.com/apache/apisix/issues/10235

**Comment by Zeeeeta** [comment]
I successfully use the ldap-auth plugin on UAT with the following config:

```
Consumer
"user_dn": "CN=uat_xxx_yyy,OU=System ID,DC=XXXX,DC=COM,DC=UAT"

Route Plugin
"base_dn": "OU=System ID,DC=XXXX,DC=COM,DC=UAT",
"uid": "CN"
```

However, when I try to use it on our PRD env, where the CN and sAMAccountName are different.
CN has space in it (XXX YY), and sAMAccountName has a dot in it instead (xxx.yy).

I am unable to authenticate no matter setting uid to CN and use "XXX YY", or setting uid to sAMAccountName and use "xxx.yy".

**Reply by moonming** [reply]
@shreemaan-abhishek please take a look, thx

**Reply by shreemaan-abhishek** [reply]
@Zeeeeta I am not much familiar with LDAP so could you please provide me with guided steps to reproduce this situation so that I can investigate and provide you a solution?

**Reply by jp-gouin** [reply]
The trim of the space is due to `obj.username = ngx.re.gsub(res[1], "\\s+", "", "jo") ` that was copy from the `basic-auth` plugin. 

Maybe it's a security thing otherwise probably not useful.
